File: topos/graphs/uast/mapper_rust.py
from __future__ import annotations

from tree_sitter import Node

from topos.graphs.uast.mapper_common import map_tree_sitter_to_uast
from topos.graphs.uast.models import UASTNode

_DECLARATION_TYPES = {
    "function_definition": "FunctionDecl",
    "class_definition": "TypeDecl",
    "struct_item": "TypeDecl",
    "enum_item": "TypeDecl",
    "impl_item": "TypeDecl",
    "trait_item": "TypeDecl",
    "function_item": "FunctionDecl",
    "method_definition": "MethodDecl",
    "lexical_declaration": "VarDecl",
    "variable_declaration": "VarDecl",
}

# Martin Abstractness classification for TypeDecl nodes. `impl_item` is
# intentionally absent: it implements an existing type rather than declaring
# a new one, so it must not be double-counted in the abstract/concrete ratio.
_TYPE_KIND = {
    "trait_item": "trait",
    "struct_item": "struct",
    "enum_item": "enum",
}

_STATEMENT_TYPES = {
    # Rust grammar exposes the control-flow primitives as expressions,
    # not statements (e.g. `if_expression`).  Both spellings map here so
    # the CFG builder picks them up uniformly across languages.
    "if_statement": "IfStmt",
    "if_expression": "IfStmt",
    "for_statement": "ForStmt",
    "for_expression": "ForStmt",
    "while_statement": "WhileStmt",
    "while_expression": "WhileStmt",
    "loop_expression": "WhileStmt",  # Rust `loop { ... }`
    "match_statement": "MatchStmt",
    "match_expression": "MatchStmt",
    "return_statement": "ReturnStmt",
    "return_expression": "ReturnStmt",
    "break_statement": "BreakStmt",
    "break_expression": "BreakStmt",
    "continue_statement": "ContinueStmt",
    "continue_expression": "ContinueStmt",
    "throw_statement": "ThrowStmt",
    "try_statement": "TryStmt",
    "expression_statement": "ExprStmt",
    "let_declaration": "VarDecl",
}

_EXPRESSION_TYPES = {
    "assignment": "AssignExpr",
    "augmented_assignment": "AssignExpr",
    "binary_expression": "BinaryExpr",
    "boolean_operator": "BinaryExpr",
    "unary_expression": "UnaryExpr",
    "call_expression": "CallExpr",
    "member_expression": "MemberExpr",
    "field_expression": "MemberExpr",
    "subscript": "MemberExpr",
}


def map_node_kind(node: Node) -> str:
    if node.type in _DECLARATION_TYPES:
        return _DECLARATION_TYPES[node.type]
    if node.type in _STATEMENT_TYPES:
        return _STATEMENT_TYPES[node.type]
    if node.type in _EXPRESSION_TYPES:
        return _EXPRESSION_TYPES[node.type]
    if node.type == "identifier":
        return "Identifier"
    if node.type.endswith("literal") or node.type in {"string", "integer", "float"}:
        return "Literal"
    if node.type in {"module", "program", "translation_unit", "source_file"}:
        return "File"
    return "Unknown"


def extract_type_attributes(node: Node) -> dict[str, object]:
    type_kind = _TYPE_KIND.get(node.type)
    return {"typeKind": type_kind} if type_kind is not None else {}


def map_rust_tree_to_uast(root: Node, file: str | None = None) -> UASTNode:
    return map_tree_sitter_to_uast(
        root=root,
        language="rust",
        map_node_kind=map_node_kind,
        file=file,
        extract_attributes=extract_type_attributes,
    )

File: topos/engine/src/graphs/uast/mapper_rust.rs
//! Rust → UAST mapper.

use std::collections::{HashMap, HashSet};

use tree_sitter::Node;

use super::mapper_common::{map_tree_sitter_to_uast, TestNodeFilter};
use super::models::{AttributeValue, UASTNode};

const CFG_TEST_MARKER: &str = "cfg(test)";

fn is_cfg_test_attribute(node: &Node, source: &[u8]) -> bool {
    node.kind() == "attribute_item"
        && node
            .utf8_text(source)
            .is_ok_and(|text| text.contains(CFG_TEST_MARKER))
}

/// Rust's [`TestNodeFilter`]: drop `#[cfg(test)]`-annotated items.
///
/// Tree-sitter-rust represents an attribute as a *preceding sibling* of
/// the item it annotates (both children of the same parent), not as a
/// descendant of that item — so this is a single forward scan over
/// `siblings`, not a per-node lookup: the `#[cfg(test)]` attribute
/// itself is dropped, and so is the item immediately following it
/// (skipping over any intervening non-`cfg(test)` attributes).
pub struct CfgTestFilter;

impl TestNodeFilter for CfgTestFilter {
    fn drop_set(&self, named_siblings: &[Node], source: &[u8]) -> HashSet<usize> {
        let mut dropped = HashSet::new();
        let mut pending_test_attr = false;
        for sibling in named_siblings {
            if sibling.kind() == "attribute_item" {
                if is_cfg_test_attribute(sibling, source) {
                    pending_test_attr = true;
                    dropped.insert(sibling.id());
                }
                continue;
            }
            if pending_test_attr {
                pending_test_attr = false;
                dropped.insert(sibling.id());
            }
        }
        dropped
    }
}

pub fn map_node_kind(kind: &str) -> &'static str {
    const NODE_KIND_TABLE: &[(&str, &str)] = &[
        ("struct_item", "TypeDecl"),
        ("enum_item", "TypeDecl"),
        ("impl_item", "TypeDecl"),
        ("trait_item", "TypeDecl"),
        ("function_item", "FunctionDecl"),
        ("let_declaration", "VarDecl"),
        ("if_expression", "IfStmt"),
        ("for_expression", "ForStmt"),
        ("while_expression", "WhileStmt"),
        ("loop_expression", "WhileStmt"),
        ("match_expression", "MatchStmt"),
        ("return_expression", "ReturnStmt"),
        ("break_expression", "BreakStmt"),
        ("continue_expression", "ContinueStmt"),
        ("expression_statement", "ExprStmt"),
        ("assignment", "AssignExpr"),
        ("augmented_assignment", "AssignExpr"),
        ("binary_expression", "BinaryExpr"),
        ("boolean_operator", "BinaryExpr"),
        ("unary_expression", "UnaryExpr"),
        ("call_expression", "CallExpr"),
        ("member_expression", "MemberExpr"),
        ("field_expression", "MemberExpr"),
        ("subscript", "MemberExpr"),
        ("identifier", "Identifier"),
        ("module", "File"),
        ("program", "File"),
        ("translation_unit", "File"),
        ("source_file", "File"),
    ];
    if let Some((_, mapped)) = NODE_KIND_TABLE.iter().find(|(k, _)| *k == kind) {
        return mapped;
    }
    if kind.ends_with("literal") || matches!(kind, "string" | "integer" | "float") {
        return "Literal";
    }
    "Unknown"
}

/// Martin Abstractness classification for `TypeDecl` nodes. `impl_item`
/// is intentionally absent: it implements an existing type rather than
/// declaring a new one, so it must not be double-counted in the
/// abstract/concrete ratio.
fn type_kind(kind: &str) -> Option<&'static str> {
    match kind {
        "trait_item" => Some("trait"),
        "struct_item" => Some("struct"),
        "enum_item" => Some("enum"),
        _ => None,
    }
}

fn extract_type_attributes(node: &Node, _source: &[u8]) -> HashMap<String, AttributeValue> {
    match type_kind(node.kind()) {
        Some(kind) => HashMap::from([(
            "typeKind".to_string(),
            AttributeValue::Str(kind.to_string()),
        )]),
        None => HashMap::new(),
    }
}

pub fn map_rust_tree_to_uast(root: Node, source: &[u8], file: Option<&str>) -> UASTNode {
    map_tree_sitter_to_uast(
        root,
        "rust",
        map_node_kind,
        source,
        file,
        Some(&CfgTestFilter),
        Some(&extract_type_attributes),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use tree_sitter::Parser;

    fn parse(source: &str) -> tree_sitter::Tree {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_rust::LANGUAGE.into())
            .unwrap();
        parser.parse(source, None).unwrap()
    }

    fn collect_kinds(node: &UASTNode, out: &mut HashSet<String>) {
        out.insert(node.native.node_kind.clone());
        for child in &node.children {
            collect_kinds(child, out);
        }
    }

    #[test]
    fn drops_cfg_test_module() {
        let source = "fn add(a: i32, b: i32) -> i32 { a + b }\n\n#[cfg(test)]\nmod tests {\n    #[test]\n    fn it_adds() {}\n}\n";
        let tree = parse(source);
        let uast = map_rust_tree_to_uast(tree.root_node(), source.as_bytes(), None);
        let mut kinds = HashSet::new();
        collect_kinds(&uast, &mut kinds);
        assert!(!kinds.contains("mod_item"));
        assert!(kinds.contains("function_item"));
    }

    #[test]
    fn keeps_non_test_attributes() {
        let source = "#[derive(Debug)]\nstruct Point { x: i32, y: i32 }\n";
        let tree = parse(source);
        let uast = map_rust_tree_to_uast(tree.root_node(), source.as_bytes(), None);
        let mut kinds = HashSet::new();
        collect_kinds(&uast, &mut kinds);
        assert!(kinds.contains("struct_item"));
    }
}
